fix: disable restored mocks

KuznetsovRoman opened the pull request:
## Proposed changes

So `Mock` has `clear` and `restore`, but it does not stop requests/responses from being processed just like active mocks.

- `request` / `match` / `continue` / ... events will still be emitted, and now there is no way to prevent it.

Example:
```ts
    it('some test', async () => {
        const mock = await browser.mock('**', {
            headers: {
                'content-type': 'text/html'
            }
        })
        mock.on('request', () => console.log('opening wdio.io page'))
        await browser.url('https://webdriver.io')
        // logs 'opening wdio.io page' once
    })
    
    it('some other test', async () => {
        const mock = await browser.mock('**', {
            headers: {
                'content-type': 'text/html'
            }
        })
        mock.on('request', () => console.log('opening wdio.io page'))
        await browser.url('https://webdriver.io')
        // logs 'opening wdio.io page' twice
    })
```
- It affects memory usage.

Example: 

```ts
it('test 0', async () => {
    const mock = await browser.mock('**');
    mock.restore();
    setInterval(() => {
        console.log(mock.calls.length);
    }, 5000);
})
Array(100).fill(0).map((_, i) => {
    it(`test ${i + 1}`, async () => {
        await browser.url(`https://youtube.com`);
    })
})
``` 
produces this log (tail), where the number comes from `console.log(mock.calls.length)`:
```
3352
[16:48:06 GMT+3] ✓ test 94 [chrome:66454ccd-cb0d-45f2-9dcd-f30994a3e0f6] - 2331ms
[16:48:09 GMT+3] ✓ test 95 [chrome:66454ccd-cb0d-45f2-9dcd-f30994a3e0f6] - 2354ms
3421
[16:48:11 GMT+3] ✓ test 96 [chrome:66454ccd-cb0d-45f2-9dcd-f30994a3e0f6] - 2456ms
[16:48:14 GMT+3] ✓ test 97 [chrome:66454ccd-cb0d-45f2-9dcd-f30994a3e0f6] - 2796ms
3491
[16:48:17 GMT+3] ✓ test 98 [chrome:66454ccd-cb0d-45f2-9dcd-f30994a3e0f6] - 2592ms
[16:48:19 GMT+3] ✓ test 99 [chrome:66454ccd-cb0d-45f2-9dcd-f30994a3e0f6] - 2456ms
3555
[16:48:24 GMT+3] ✓ test 100 [chrome:c8ffdf93-165f-4377-9f09-16a9de8e9247] - 3174ms
```
so, it looks like those calls are being saved, and if each test has a mock, then 100-th test will be stored 100 times.

Example:

```ts
Array(100).fill(0).map((_, i) => {
    it(`test ${i + 1}`, async () => {
        const mock = await browser.mock('**');
        await browser.url(`https://youtube.com`);
        mock.restore(); // replace with mock.remove to fix exponential memory leak
    })
})
```
uses 3.5GB ram:
<img width="1072" alt="image" src="https://github.com/webdriverio/webdriverio/assets/54893992/d7729ff7-00b9-4482-9823-7e48ff390679">

- It affects performance (not much, really, but still): with just `mock.restore` at the end of test, 100-th test takes 1171 ms, and with `mock.remove` - 810ms (on simple test: `mock = browser.mock('**'); browser.url('localhost:3000'); mock.remove`, where localhost:3000 serves react app template, generated with create-react-app). `mock.remove` also disables fetch domain, if there are no other mocks. I assume, fetch domain has some effect on performance.

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have added proper type definitions for new commands (if appropriate)

### Reviewers: @webdriverio/project-committers

christian-bromann commented on packages/webdriverio/src/commands/browser/mock.ts:
```diff
@@ -172,6 +177,25 @@ export async function mock (
     const networkInterception = new NetworkInterception(url, filterOptions, browser)
     SESSION_MOCKS[handle].add(networkInterception as Interception)
 
+    networkInterception.remove = () => {
```
I agree, this is not the right place to define this command in and we should find ways to put it in `NetworkInterception`.

> you need networkInterception in order to remove it from set

If you define this method in the `NetworkInterception` class it is accessible through `this`.

> you need SESSION_MOCKS, to know, where to delete from

`SESSION_MOCKS` then needs to be defined somewhere where `mock.ts` and the file where `NetworkInterception` is defined can import it from without having circular dependencies.

> you need handle, to know where to exactly delete from

make `remove` async and get it via `const handle = await browser.getWindowHandle()`

> you need this client: CDPSession to disable fetch domain,

You have access to the `browser` variable so you can get that variable too.

christian-bromann replied:
The method `mockRestore` is meant to do what you want to do with `remove` .. can we just make sure the restore method ensures that nor memory is hogged?

KuznetsovRoman replied:
> This is an issue that should be addressed but I am not happy with the current approach

That's why i converted it to draft: i found a better way to do it

KuznetsovRoman replied:
> You have access to the browser variable so you can get that variable too

i tried, and that's what i discovered: you can get puppeteer through browser, but not `client: CDPSession`, because it would be different client, and we need original one.

KuznetsovRoman replied:
> The method `mockRestore` is meant to do what you want to do with `remove` .. can we just make sure the restore method ensures that nor memory is hogged?

There is also a moment to consider. There are two options:
- `restore` also sends `Fetch.disable`, if there are no other mocks (if `restore` means "i dont need this anymore". In this case we would have to enable fetch domain again after some `respond` / `abort`)
- `restore` does not disable fetch domain, even if there are no other mocks. (if you plan to use `respond` / `abort` after it was restored. In this case we could just ensure this mock presents in `SESSION_MOCKS`)

First option also makes `restore` async for  devtools and breaks existing "respond / abort after restore" cases, so i think about implementing second approach, but at first i need to clarify: what happens if you try to "respond / abort after restore" with saucelabs? Will it work just like mock wasn't restored?

christian-bromann replied:
We have designed the API to be as close as possible to the [Jest Mock API](https://jestjs.io/docs/mock-function-api#mockfnmockrestore) so I would say the first should be more appropriate. I feel honestly less worried about the Sauce Labs integration. I was working there at a time I build this and don't have any usage numbers. At the end of the day this API should support what will be supported in WebDriver Bidi because Sauce eventually has to support that.

christian-bromann reviewed (commented):
This is an issue that should be addressed but I am not happy with the current approach.

KuznetsovRoman commented on packages/webdriverio/src/commands/browser/mock.ts:
```diff
@@ -4,8 +4,10 @@ import WebDriverNetworkInterception from '../../utils/interception/webdriver.js'
+import type { CDPSession } from 'puppeteer-core/lib/esm/puppeteer/common/Connection.js'
 
 export const SESSION_MOCKS: Record<string, Set<Interception>> = {}
+export const CDP_SESSIONS: Record<string, CDPSession> = {}
```
`const client = await page.target().createCDPSession()` creates new CDPSession, but we need the original one to disable fetch domain. Storing `map<handle, CDPSession>` allows us to restore mocks in any order

KuznetsovRoman commented on packages/webdriverio/src/utils/interception/devtools.ts:
```diff
@@ -221,9 +224,25 @@ export default class DevtoolsInterception extends Interception {
      * removes any mocked return values or implementations.
      */
-    restore () {
+    async restore (sessionMocks = SESSION_MOCKS, cdpSessions = CDP_SESSIONS) {
```
this is for unit tests

KuznetsovRoman commented on packages/webdriverio/src/utils/interception/devtools.ts:
```diff
@@ -262,6 +284,12 @@ export default class DevtoolsInterception extends Interception {
         this.abort(errorReason, false)
     }
+
+    private ensureNotRestored() {
+        if (this.restored) {
+            throw new Error('This can\'t be done on restored mock')
+        }
+    }
```
If somebody has tests, where `abort` / `respond` is called after `restore`, currently it would work just like creating the new one with mocked response.

After this commit, `abort` / `respond` after `restore` just won't do anything, so i added this check to let people know, that you can't do that now.

Also, should i include `BREAKING CHANGE` in commit message footer? That could be considered as breaking change.

KuznetsovRoman reviewed (commented):
Now `mock.restore()` acts like reversed `browser.mock()`.

KuznetsovRoman commented:
Also added a commit to handle error responses.

Currently, event is considered a request event, if it has no status code, but some non-request events also dont have it.

Event example:

```ts
{
  requestId: 'interception-job-3.0',
  request: {
    url: 'https://foo-bar-baz-qux.com/image.jpeg', // when host is not resolved
    method: 'GET',
    headers: { ... },
    initialPriority: 'Low',
    referrerPolicy: 'strict-origin-when-cross-origin'
  },
  frameId: '1A8B140A1A0109D1F6C38483096CB34F',
  resourceType: 'Image',
  responseErrorReason: 'NameNotResolved',
  networkId: '7803.5'
}
```

I fixed it, so those events would be handled as not-requests, and we would see them.

For example:
```ts
mock.on("match", match => {
    console.log(match.statusCode); // will prints "0", if there is no such host. Currently, we just wont have that event.
})
```

0 is also a falsy value, so we could use `!match.statusCode`

christian-bromann commented on packages/webdriverio/src/utils/interception/devtools.ts:
```diff
@@ -220,10 +224,27 @@ export default class DevtoolsInterception extends Interception {
+        }
+
+        log.trace(`Disabling fetch domain for ${handle}`)
+        return cdpSessions[handle].send('Fetch.disable')
```
This would restore mocking for all mocks right? Can you verify that you can create two mocks within the same session and have only one restored?

KuznetsovRoman replied:
Yup. Fixed in https://github.com/webdriverio/webdriverio/pull/10413/commits/ee4f27650c015bb7ddc2a8a1eff56542fa2edcb3

Also changed type [here](https://github.com/webdriverio/webdriverio/pull/10413/commits/ee4f27650c015bb7ddc2a8a1eff56542fa2edcb3#diff-0897b632afeab5d0ee660ee9e1ad90fade9350a431d9090571e4c934a7ea8ec9L14-R14)
and [used await instead of return here](https://github.com/webdriverio/webdriverio/pull/10413/commits/ee4f27650c015bb7ddc2a8a1eff56542fa2edcb3#diff-81b3fe6cd1e82bc9b5323f327cb91b6fba1a061ab6509f3b451676517a1ea598L50-R51), because `Promise<Promise<void>>` and `Promise<void>` are not the same thing

christian-bromann reviewed (commented):
Could it be that what you are trying to do is more close towards what a [`restoreAllMocks`](https://jestjs.io/docs/jest-object#jestrestoreallmocks) would do?

christian-bromann commented:
@KuznetsovRoman it seems like some unit tests are failing 🤔

KuznetsovRoman commented:
> @KuznetsovRoman it seems like some unit tests are failing thinking

Yes, it's the one, checking webdriver's mock "restore" method return value. I changed "return" to "await" there, so it now returns undefined, instead of "{}". Do we need this return value, or why do we even need this check in the first place?

christian-bromann commented:
> Do we need this return value, or why do we even need this check in the first place?

We don't need the return value, just have the expected value be `undefined` then. The check is good so we make sure whatever we call works and doesn't throw, due to non existence for example

KuznetsovRoman commented:
Should be fixed

christian-bromann reviewed (approved):
Thank you, LGTM 👍
